### ISSUE TITLE
Simulation with tilted gravity and background buoyancy increases in total energy

### ISSUE THREAD

**Issue by tomchor** [CliMA/Oceananigans.jl#3036]
The example below creates a simulation with a rotated gravity vector, a background buoyancy field that's just a stratified background with linear stratification and it starts everything off with some resolved noise:

```julia
using Oceananigans
using Oceananigans.Units
using Oceananigans.Grids: min_Δx, min_Δz

θ_rad = 0.15
N²∞ = 5e-6
grid = RectilinearGrid(CPU(), topology=(Periodic, Flat, Bounded),
                       size=(50, 30), extent = (200, 80))

@inline b_bg(x, y, z, t, p) = p.N²∞ * (x * p.g1 +  z * p.g3)
B_field = BackgroundField(b_bg, parameters=(; N²∞, g1 = sin(θ_rad), g3 = cos(θ_rad)))

model = NonhydrostaticModel(grid = grid, timestepper = :RungeKutta3,
                            advection = UpwindBiasedFifthOrder(),
                            buoyancy = Buoyancy(model=BuoyancyTracer(), gravity_unit_vector=-[sin(θ_rad), 0, cos(θ_rad)]),
                            tracers = :b,
                            background_fields = (; b=B_field),
                            )

amp = 1e-3
xₚ = grid.Lx*rand(5); zₚ = -grid.Lz*rand(5) # random locations for 5 gaussians
σx = 4min_Δx(grid);   σz = 4min_Δz(grid) # spread of Gaussians
resolved_noise(x, y, z) = sum(@. amp * exp(-(x-xₚ)^2/σx^2 -(z-zₚ)^2/σz^2))
set!(model, u=resolved_noise, w=resolved_noise)

u, v, w = model.velocities
b = model.tracers.b

using Statistics: mean
u.data.parent .-= mean(u)
w.data.parent .-= mean(w)

simulation = Simulation(model, Δt = 0.5 * min_Δz(grid) / maximum(abs, v),
                        stop_time = 3days)

wizard = TimeStepWizard(max_change=1.01, cfl=0.2, diffusive_cfl=0.1, max_Δt=0.2/√N²∞)
simulation.callbacks[:wizard] = Callback(wizard, IterationInterval(2))

KE = Average((u^2 + w^2) / 2)
PE = Average(b^2/(2*N²∞))
outputs = (; u, w, b, KE, PE)

simulation.output_writers[:writer] = NetCDFOutputWriter(model, outputs;
                                                        filename = "mwe.nc",
                                                        schedule = TimeInterval(1hour),
                                                        with_halos = false,
                                                        overwrite_existing = true)
run!(simulation)

using Rasters, Plots
fpath_xz1 = simulation.output_writers[:writer].filepath
ds_xz1 = RasterStack(fpath_xz1)
u_lims = (-amp, +amp)

times = dims(ds_xz1, :Ti)
anim = @animate for frame in 1:length(times)
    p1 = heatmap(ds_xz1[:u][Ti=frame, yC=1]', clim=u_lims, c=:balance, title="u")
    p2 = heatmap(ds_xz1[:w][Ti=frame, yC=1]', clim=u_lims, c=:balance, title="w")
    p3 = heatmap(ds_xz1[:b][Ti=frame, yC=1]', title="b")
    p4 = plot(ds_xz1[:KE])
    plot!(p4, ds_xz1[:PE])
    vline!(p4, [times[frame]], label="time", linestyle=:dash)

    fullplot = plot(p1, p2, p3, p4, layout = Plots.grid(2, 2), size=(1000, 600))
end

mp4(anim, "mwe.mp4", fps=16)

```

I expected this simulation to slowly lose or approximately conserve total energy. Instead I get these grid-scale noises popping up and a build up of kinetic and potential energy seemingly out of nowhere:

https://user-images.githubusercontent.com/13205162/228424859-905f3dd6-b167-440b-b073-af09008f4f04.mp4

Notes: 

- Even if I add a background diffusivity with a reasonable value, the energy buildup still happens
- The energy buildup does not happen without a background buoyancy
- When `gravity_unit_vector == NegativeZDirection()` the KE stays approximately the same, but the PE increases. Although that might be due to the boundary conditions, I'm not sure at this point.
- When I start this with a larger velocity amplitude the small scale noise becomes less evident, but there's still an increase in the total energy (KE and PE) of the flow

Any guesses as to what might be causing this?

**Comment by navidcy** [comment]
OK, perhaps I'm a bit oblivious to physics of tilted gravity etc. Are they crucially important for this to happen?

Also, is there an energy integral associated with these dynamics when there is a fixed imposed buoyancy? Usually systems with imposed mean flow don't have an energy integral and this is a side-effect of the system not being energetically closed since something is continuously putting energy in to keep that mean flow fixed. I'm not sure if this is the case here, but here's an example:

$$
\begin{align}
\partial_t u + u\partial_x u + v\partial_y u - f v = ν \nabla^2 u \\
\partial_t v + u\partial_x v + v\partial_y v + f u = ν \nabla^2 v
\end{align}
$$

implies that 

$$
\begin{align}
\frac{d}{dt} \int \frac1{2} (u^2+v^2) \mathrm{d}^2\boldsymbol{x} &= \int (u \partial_t u + v \partial_t v)  \mathrm{d}^2\boldsymbol{x} \\
&= - \int (u^2 \partial_x u + u v \partial_y u  + u v \partial_x v + v^2 \partial_y v)  \mathrm{d}^2\boldsymbol{x} +  ν \int (u \nabla^2 u + v \nabla^2 v) \mathrm{d}^2\boldsymbol{x}  \\
&= - \int [u \partial_x ( u^2+v^2)/2 + v \partial_y ( u^2+v^2)/2 ]  \mathrm{d}^2\boldsymbol{x} +  ν \int (u \nabla^2 u + v \nabla^2 v) \mathrm{d}^2\boldsymbol{x}  \\
&= - \int \boldsymbol{\nabla \cdot} \left[ \boldsymbol{u} \ \frac1{2}( u^2+v^2) \right]  \mathrm{d}^2\boldsymbol{x} - ν \int | \boldsymbol{\nabla} \boldsymbol{u}|^2 \\
&= - ν \int | \boldsymbol{\nabla} \boldsymbol{u}|^2   \mathrm{d}^2\boldsymbol{x}  < 0
\end{align}
$$

But, if there is an imposed zonal flow $U(y)$ then:

$$
\begin{align}
\partial_t u + U \partial_x u + u\partial_x u + v \partial_y U + v \partial_y u - f v = ν \nabla^2 u \\
\partial_t v + U\partial_x v  + u\partial_x v + v\partial_y v + f u = ν \nabla^2 v
\end{align}
$$

and then 

$$
\begin{align}
\frac{d}{dt} \int \frac1{2} (u^2+v^2) \mathrm{d}^2\boldsymbol{x} &= \int (u \partial_t u + v \partial_t v)  \mathrm{d}^2\boldsymbol{x} \\
&= - \int (u U \partial_x u + u v \partial_y U - U v \partial_x v)  \mathrm{d}^2\boldsymbol{x}  - \int (u^2 \partial_x u + u v \partial_y u  + u v \partial_x v + v^2 \partial_y v)  \mathrm{d}^2\boldsymbol{x} +  ν \int (u \nabla^2 u + v \nabla^2 v) \mathrm{d}^2\boldsymbol{x}  \\
&= - \int (u v \partial_y U)  \mathrm{d}^2\boldsymbol{x}  - ν \int | \boldsymbol{\nabla} \boldsymbol{u}|^2   \mathrm{d}^2\boldsymbol{x} 
\end{align}
$$

which has not a definite sign.

**Comment by navidcy** [comment]
the initial condition is not necessarily incompressible -- did you try `enforce_incompressibility = true`?

**Comment by tomchor** [comment]
> the initial condition is not necessarily incompressible -- did you try `enforce_incompressibility = true`?

Incompressibility in `set!()` is enforced by default: https://github.com/CliMA/Oceananigans.jl/blob/bcc34f07b3f949ea6fb34c7814f4b856d24924c2/src/Models/NonhydrostaticModels/set_nonhydrostatic_model.jl#L32

But I did explicitly include the flag, just in case, and the results indeed were the same.

**Comment by simone-silvestri** [comment]
Just to clarify a point here. 
The background field here does not participate in the hydrostatic pressure, it only adds an advective term to the tracer equation which reads $\nabla \cdot (boldsymbol{u} B)$. So in practice, you are solving

$$\partial_t boldsymbol{u} + (boldsymbol{u} \cdot \nabla) boldsymbol{u}  = - \nabla p + \widehat{g} b + \mathcal{D}_i(boldsymbol{u})$$
$$\partial_t b + $\nabla \cdot (boldsymbol{u} b)$ + $\nabla \cdot (boldsymbol{u} B)$ = \mathcal{D}_i(b) +\mathcal{D}_i(boldsymbol{B}) $$
where `\mathcal{D}_i` is the implicit diffusion coming from 5th order upwind

I guess I am confused about why would you add that advection term to the equation, what does it add to your physics?

**Comment by tomchor** [comment]
> Just to clarify a point here. The background field here does not participate in the hydrostatic pressure, it only adds an advective term to the tracer equation which reads ∇⋅(uB). So in practice, you are solving
> 
> ∂tu+(u⋅∇)u=−∇p+g^b+Di(u) ∂tb+∇⋅(ub)+∇⋅(uB)=Di(b)+Di(B) where Di is the implicit diffusion coming from 5th order upwind

Yup, and just to clarify the physics behind this set-up, it is necessary when simulating things like doubly-periodic tilted boundary layers in a stratified environment (such as our [tilted bottom example](https://clima.github.io/OceananigansDocumentation/stable/generated/tilted_bottom_boundary_layer/)).

@simone-silvestri not to sidetrack from the issue but isn't $\mathcal D_i(B)$ ignored in the evolution equation? I got that impression from reading the docs [here](https://clima.github.io/OceananigansDocumentation/stable/model_setup/background_fields/) and [here](https://clima.github.io/OceananigansDocumentation/stable/physics/nonhydrostatic_model/).

**Comment by simone-silvestri** [comment]
$\mathcal{D}_i(B)$ is implicit in your advection if you use upwinding. 
Now we have to understand if we actually need $\mathcal{D}_i(B)$ or not in the $b$ evolution equation.

**Comment by tomchor** [comment]
@simone-silvestri I was able to try this simulation using the same advection scheme for all velocities and fields using the branch in https://github.com/CliMA/Oceananigans.jl/pull/3027. More precisely I resolved the same simulation with only one call to `div_Uc` using `SumOfArrays(velocities)` and `SumOfArrays(tracers)`, which I think is the correct way to test this, and the results were the same.

Furthermore, I also ran this without doing hydrostatic pressure separation using [this branch](https://github.com/CliMA/Oceananigans.jl/tree/tc/nhpressure2) and the results were pretty disastrous:

https://user-images.githubusercontent.com/13205162/233477175-159e2c66-63dc-44aa-80a8-142c06d04443.mp4

That said, I think in this last animation the noise emerges from the top and propagates to the bottom, while in all my other attempts the noise always appears to emerge at the bottom and propagate upwards. I wonder if that's a clue as to what's happening.

**Comment by simone-silvestri** [comment]
that looks very weird! But it is structured in some sense, it looks like the initial condition has some inbalances that generate noise which gets cleaned up by numerical diffusion. Interesting also that buoyancy extrema are continuously increasing in magnitude. If this is not happening with z-directed gravity, it might really be that there is some bug/incorrectness in the buoyancy model. A way to test this would be to recreate the same setup with z-directed gravity and use an immersed boundary to represent the tilt. If that case behaves as expected we know where to focus.

**Comment by glwagner** [comment]
Have you tried adding background diffusion? We found in another issue that nearly linear dynamics can be noisy even with upwinding (when the velocities are small, the diffusion added by upwinded advection can be negligible). Laplacian or hyperdiffusion may help control noise.

**Comment by tomchor** [comment]
> Have you tried adding background diffusion? We found in another issue that nearly linear dynamics can be noisy even with upwinding (when the velocities are small, the diffusion added by upwinded advection can be negligible). Laplacian or hyperdiffusion may help control noise.

Yes I did. The issue still persists with all advection schemes I've tried so far and with any "reasonable" value of molecular diffusion. I say reasonable because if I keep cranking up the diffusivity, at some point the energy accumulation will stop, but so will most of the interesting dynamics that one might be interested in.

As an example, here's how the simulation looks like with `ScalarDiffusivity(ν=1e-4, κ=1e-4)` and `θ_rad = 0.15`:

https://user-images.githubusercontent.com/13205162/233661938-73c71f1c-2a9e-4013-ab3d-e23a29e0d5b6.mp4

I only get a decrease in energy over time if I use a viscosity/diffusivity of around 1e-3, which I consider to be pretty high.

**Comment by tomchor** [comment]
> that looks very weird! But it is structured in some sense, it looks like the initial condition has some inbalances that generate noise which gets cleaned up by numerical diffusion. Interesting also that buoyancy extrema are continuously increasing in magnitude. If this is not happening with z-directed gravity, it might really be that there is some bug/incorrectness in the buoyancy model. A way to test this would be to recreate the same setup with z-directed gravity and use an immersed boundary to represent the tilt. If that case behaves as expected we know where to focus.

@simone-silvestri I'm not sure if that'd be an apples-to-apples comparison. We'd be comparing a domain with smooth tilted boundaries (using the rotated buoyancy) against a domain with jagged tilted boundaries (using our grid-fitted IB). So I'm not sure what we could learn there (please lmk if I'm missing something here). That said, I think it's worth the shot so I might try to implement that soon and see what happens.

**Comment by simone-silvestri** [comment]
It would for sure introduce noise because of the jagged bottom, maybe try cranking up a bit the resolution to have less bathymetry imprinting. But, if the energy does not increase we found our culprit.
Anyways, I don't think it's an advection scheme problem. It might be something more fundamental about this setup or about tilted gravity per se.

**Comment by tomchor** [comment]
In the meantime, here's something interesting that I found out: the dynamics seem to be symmetrical with respect to the tilt. For example, the animations for `θ_rad=+π/2` and `θ_rad=-π/2` below look virtually the same, and I expected them to be flipped images of each other:

https://user-images.githubusercontent.com/13205162/233666649-fd267937-cfcb-43f1-9784-113cbbdbeccf.mp4

https://user-images.githubusercontent.com/13205162/233666634-25613eb1-ca44-4f3e-a799-e31d98a8201c.mp4

The only difference you can see in these simulations is the buoyancy distribution, whose z-mean is flipped (because the vertical direction point at opposite directions in these two simulations). I think this is a clue.

PS: In these simulations I'm using a bounded domain (with `NoFlux` in all directions) in both x and z since the tilt is 90 degrees.

**Comment by glwagner** [comment]
> Yes I did. The issue still persists with all advection schemes I've tried so far and with any "reasonable" value of molecular diffusion.

Maybe I wasn't clear, the point is that if the dynamics are linear or small-amplitude then advection doesn't play a role; it wouldn't matter what advection scheme you use. The term is small.

Does hyperviscosity allow stable simulations with non-trivial dynamics?

**Comment by glwagner** [comment]
I think I found a bug. See #3079.

**Comment by tomchor** [comment]
> > Yes I did. The issue still persists with all advection schemes I've tried so far and with any "reasonable" value of molecular diffusion.
> 
> Maybe I wasn't clear, the point is that if the dynamics are linear or small-amplitude then advection doesn't play a role; it wouldn't matter what advection scheme you use. The term is small.

I see. Agreed.

> Does hyperviscosity / hyperdiffusion allow stable simulations with non-trivial dynamics?

I haven't tried it. Possibly. But the simulations I'm using this set-up for are LES. Assuming adding hyperviscosity does do what you're suggesting, I'm not sure how "correct" it is to use that along with a large-eddy simulation closure. (I think here are arguments to be made for both sides.)

That said, it's worth the shot. Although tbh I have no idea what a reasonable value for hyperviscosity is since I've never used it :joy: . What would be a fair value to include in this MWE?

**Comment by glwagner** [comment]
Hyperviscosity _is_ an LES closure. It's not just a great one because it requires tuning (ie you have to select the hyperviscosity, and the hyperviscosity doesn't change).

You can use a scaling argument to find the hyperviscosity:

$$ \nu_4 = \Delta^4 / \tau $$

where $\Delta$ is a length scale (typically the grid spacing) and $\tau$ is the time-scale over which you'd like to damp grid scale motions. The longer the better. Perhaps you can estimate a time-scale in terms of the buoyancy frequency (eg to damp grid scale buoyancy oscillations).

**Comment by tomchor** [comment]
This is weird, while I do agree with you that https://github.com/CliMA/Oceananigans.jl/pull/3079 seems like the right way to go, it seems it makes matters worse...

This is the animation using [this code](https://pastebin.com/FRmZ9Hhe) on main (it's just an up-to-date version of the code in the top comment):

https://user-images.githubusercontent.com/13205162/233679645-209cbdf3-762e-47ed-a4dd-175a8061ec6a.mp4

And this is the output of the same code using the branch on https://github.com/CliMA/Oceananigans.jl/pull/3079:

https://user-images.githubusercontent.com/13205162/233679749-9c1164af-9396-4f91-bfcf-44c2791cf9a9.mp4

So it seems the bugfix is actually worsening the results...

**Comment by glwagner** [comment]
The above post seems to contradict what you said on the PR, can you clarify?

**Comment by tomchor** [comment]
> The above post seems to contradict what you said on the PR, can you clarify?

My post here happened was before you fixed the perturbation buoyancy reconstruction on the pressure calculation and before I tried your bugfix without a pressure separation (which made the correction in the pressure calculation unnecessary).

https://github.com/CliMA/Oceananigans.jl/pull/3079 as it currently is fixes everything for me (that I tried so far).
